### ISSUE TITLE
Regression in `prop-types` rule in v5.1.1

### ISSUE THREAD

**Issue by EvNaverniouk** [yannickcr/eslint-plugin-react#591]
After upgrading from 5.0.1 to 5.1.1 I noticed a bug with the `prop-types` rule. It now incorrectly detects destructured method params as missing props. See example below:

```javascript
export default (ComposedComponent) => class Something extends SomeOtherComponent {
    someMethod = ({width}) => {
        this.setState({width});
    }
}
```

Will result in an incorrect failure of: `react/prop-types: 'width' is missing in props validation`.

What's strange is that all of the examples below do NOT fail:

```javascript
// Removed destructured function param
export default (ComposedComponent) => class Something extends SomeOtherComponent {
    someMethod = (width) => {
        this.setState({width});
    }
}

// Kept destructured function param but extend the class via `Component` instead
export default (ComposedComponent) => class Something extends Component {
    someMethod = ({width}) => {
        this.setState({width});
    }
}

// Kept destructured function param, and the custom class extend,
// but removed the composed function wrapper
export default class Something extends SomeOtherComponent {
    someMethod = ({width}) => {
        this.setState({width});
    }
}
```

Seems like the combination of a destructured function param, a HOC composed function wrapper and extending from a non-`Component` class causes a failure of the rule.

**Comment by strawbrary** [comment]
Not sure if this is the same issue, but I've been having a similar issue with decorators and destructured params when wrapped with an array and/or object.

The following results in `error  'store' is missing in props validation  react/prop-types`.
```javascript
@asyncConnect([{
  promise: ({ store: { dispatch } }) => {
    return dispatch(fetchUsers());
  },
}])
```

The following do not trigger errors.
```javascript
// No error if function isn't wrapped by an object and/or array
@asyncConnect(
  ({ store: { dispatch } }) => {
    return dispatch(fetchUsers());
  }
)

// No error after moving destructuring into function body
@asyncConnect([{
  promise: (params) => {
    const { store: { dispatch } } = params;
    return dispatch(fetchUsers());
  },
}])
```

**Comment by yannickcr** [comment]
After some tests the regression seems to come from https://github.com/yannickcr/eslint-plugin-react/commit/4e1bcc0020834f06801a6987043e085658e72bd7 so it should be here since v4.1.0
